8299546: C2: MulLNode::mul_ring() wrongly returns bottom type due to casting errors with large numbers

chhagedorn opened the pull request:
The current logic in `MulLNode::mul_ring()` casts all `jlong` values of the involved type ranges of a multiplication to `double` in order to catch overflows when multiplying the two type ranges. This works fine for values in the `jlong` range that are not larger than 2<sup>53</sup> or lower than -2<sup>53</sup>. For numbers outside that range, we could experience precision errors because these numbers cannot be represented precisely due to the nature of how doubles are represented with a 52 bit mantissa. For example, the number 2<sup>53</sup> and 2<sup>53</sup> + 1 both have the same `double` representation of 2<sup>53</sup>.

In `MulLNode::mul_ring()`, we could do a multiplication with a `lo` or `hi` value of a type that is larger than  2<sup>53</sup> (or smaller than -2<sup>53</sup>). In this case, we might get a different result compared to doing the same multiplication with `jlong` values (even though there is no overflow/underflow). As a result, we return `TypeLong::LONG` (bottom type) and missed an optimization opportunity (e.g. folding an `If` when using the `MulL` node in the condition etc.).

This was caught by the new verification code added to CCP in [JDK-8257197](https://bugs.openjdk.org/browse/JDK-8257197) which checks that after CCP, we should not get a different type anymore when calling `Value()` on a node. In the found fuzzer testcase, we run into the precision problem described above for a `MulL` node and set the type to bottom during CCP (even though there was no actual overflow). Since the type is bottom, we do not re-add the node to the CCP worklist because the premise is that types only go from top to bottom during CCP. Afterwards, an input type of the `MulL` node is updated again in such a way that the previously imprecise `double` multiplication in `mul_ring()` is now exact (by coincidence). We then hit the "missed optimization opportunity" assert added by JDK-8257197.

To fix this problem, I suggest to switch from a `jlong` - > `double` multiplication overflow check to an overflow check without casting. I've used the idea that `x = a * b` is the same as `b = x / a` (for `a != 0` and `!(a = -1 && b = MIN_VALUE)`)  which is also applied in `Math.multiplyExact()`: https://github.com/openjdk/jdk/blob/66db0bb6a15310e4e60ff1e33d40e03c52c4eca8/src/java.base/share/classes/java/lang/Math.java#L1022-L1036

The code of `MulLNode::mul_ring()` is almost identical to `MulINode::mul_ring()`. I've refactored that into a template class in order to share the code and simplified the overflow checking by using `MIN/MAX4` instead of using nested `if/else` statements.

Thanks,
Christian

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299546](https://bugs.openjdk.org/browse/JDK-8299546): C2: MulLNode::mul_ring() wrongly returns bottom type due to casting errors with large numbers


### Reviewers
 * [Igor Veresov](https://openjdk.org/census#iveresov) (@veresov - **Reviewer**) ⚠️ Review applies to [0453789f](https://git.openjdk.org/jdk/pull/11907/files/0453789f65d9b3c755116184f912d65ba20740a8)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer) ⚠️ Review applies to [0453789f](https://git.openjdk.org/jdk/pull/11907/files/0453789f65d9b3c755116184f912d65ba20740a8)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11907/head:pull/11907` \
`$ git checkout pull/11907`

Update a local copy of the PR: \
`$ git checkout pull/11907` \
`$ git pull https://git.openjdk.org/jdk pull/11907/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11907`

View PR using the GUI difftool: \
`$ git pr show -t 11907`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11907.diff">https://git.openjdk.org/jdk/pull/11907.diff</a>

</details>

vnkozlov commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -281,44 +281,76 @@ Node *MulINode::Ideal(PhaseGVN *phase, bool can_reshape) {
+  // x / a does not underflow/overflow since |x / a| <= |x|. If a * b does underflow/overflow, then we'll get a different
+  // result compared to a * b. Special case MIN_VALUE * -1 whose result is MIN_VALUE.
+  bool does_overflow(const NativeType a, const NativeType b) const {
+    NativeType x = java_multiply(a, b);
```
Should we check for 0 values to `return false;` immediately before all calculations?
Also may be path results of `java_multiply()` to does_overflow(). Otherwise you execute it twice.

chhagedorn replied:
I've updated my patch with your suggestion.

merykitty commented:
I would suggest we do a 128-bit multiplication or a 64-bit high multiplication instead. This will help constant folding of `MulHiNode`s and also `MulLNode` even if the multiplication overflows. Also, a high multiplication is likely to be cheaper than 4 division.

Thanks.

merykitty commented:
@chhagedorn I think we can do similar to what our core library is doing

https://github.com/openjdk/jdk/blob/8b0133f2760f67cd968528c041a600408cc26978/src/java.base/share/classes/java/lang/Math.java#L1399

This helps in the sense that we currently do not have a 64-bit high multiplication inside the VM, so having one makes adding a more effective `MulHiNode::Value` become trivial. Regarding `MulLNode::Value`, even if the multiplication overflows, I think we can still reason about the value of the product if all the high parts are the same. The same goes for `MulINode::Value` where we can cast them to `jlong` instead of `double`.

Thanks.

chhagedorn commented:
I think I understand now what you mean, thanks for the explanation. That's a very interesting idea. I will try it out and get back with an updated patch proposal and some tests where this can be beneficial.

Thanks,
Christian

rose00 commented:
Today's processors support 64-to-128-bit multiply in just a few cycles.  This is a useful operation here and elsewhere.  We should bite the bullet and arrange to make it available in HotSpot.  It would make this particular problem a little simpler to solve.

Here is a good external example of it being done well in fast, portable code:

https://github.com/Cyan4973/xxHash/blob/8e5fdcbe70687573265b7154515567ee7ca0645c/xxh3.h#L294

Here is an example (not to be followed literally) of code that has recently worked well for me on both x86 and ARM:

```
typedef __int128 uint128_t;
inline uint128_t make_uint128(uint64_t hi, uint64_t lo) {
  uint128_t hi128 = hi, lo128 = lo;
  if ((hi128 | lo128) >> 64)  abort();  // please no sign extension bugs
  return (hi128 << 64) | lo128;
}
inline void take_uint128(uint64_t& hi_out, uint64_t& lo_out, uint128_t x) {
  hi_out = (uint64_t)(x >> 64);
  lo_out = (uint64_t)(x >> 0);
}
inline void full_multiply(uint64_t& hi_out, uint64_t& lo_out, uint64_t x, uint64_t y) {
  take_uint128(hi_out, lo_out, make_uint128(0, x) * make_uint128(0, y));
}
```

There should be both signed and unsigned versions of this or a similar primitive.

I suggest making the two-output thing (full-multiply, signed or unsigned) to be the primitive for HotSpot, not the mul-hi intrinsic that Java has (high-half of full-multiply, signed or unsigned).  Java tilts strongly towards one-output primitives because we don't have Project Valhalla yet, but C++ has no such restriction.

vnkozlov reviewed (approved):
Looks good.

chhagedorn commented:
I've updated the algorithm to implement @merykitty's idea to get a more precise type if all the cross products overflowed/underflowed the same number of times. I've included a lot of tests to make sure it works properly. For the implementation, I added `multiply_high_signed()` to `globalDefinitions.hpp`. I had first tried to make it work with an unsigned version but that somehow did not as some tests kept failing. I've left the unsigned `multiply_high_unsigned()` in `globalDefinitions.hpp` for future use. But I could also remove it again as it is currently unused.

As the C++ standard does not define the behavior of shifting a signed integer to the right and leaves it up to the compiler to decide, I've used a portable version. Even though, I think most compilers will use an arithmetic shift but can we be sure about that? There is this comment in the code that assumes that this is always the case:
https://github.com/openjdk/jdk/blob/b2d3622115ce0b4c0647c7b79f28c075dfcdebbc/src/hotspot/share/utilities/globalDefinitions.hpp#L1230-L1232

Thanks,
Christian

vnkozlov reviewed (approved):
Latest update looks reasonable.

chhagedorn commented:
Thanks Vladimir for reviewing it again!

veresov commented:
Looks good to me too.

chhagedorn commented:
Thanks Igor for reviewing it again!

merykitty commented on src/hotspot/share/utilities/globalDefinitions.hpp:
```diff
@@ -1155,6 +1155,44 @@ inline intx byte_size(void* from, void* to) {
+// Right shifts with signed integers are compiler implementation specific according to the C++ standard.
+// Use a portable version instead.
+inline int64_t shift_right_arithmetic(int64_t value, uint8_t shift_amount) {
+  return value < 0 ? (int64_t)(~(~(uint64_t)value >> shift_amount)) : (int64_t)((uint64_t)value >> shift_amount);
```
I believe we have `java_shift_right` already, it assumes signed extension, however.

https://github.com/openjdk/jdk/blob/master/src/hotspot/share/utilities/globalDefinitions.hpp#L1240

chhagedorn replied:
Yes, I've seen that - was just not too sure about using it as the C++ standard leaves it up to the compiler to decide (most compilers will probably just sign extend). However, I see that we have already been using `java_shift_right`. So it might be fine to just use that as well.

vnkozlov replied:
We need @kbarrett opinion on C++ code for this case.
I prefer to use already defined functions if they work to avoid duplication.

chhagedorn replied:
@kimbarrett 

Looks like we've initially pinged the wrong Kim Barrett :-)

kimbarrett replied:
C++14 5.8/3 In the description of "E1 >> E2" it says "If E1 has a signed type
and a negative value, the resulting value is implementation-defined."

However, C++20 7.6.7/3 further defines integral arithmetic, as part of
requiring two's-complement behavior.
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0907r3.html
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1236r1.html
The corresponding C++20 text is "Right-shift on signed integral types is an
arithmetic right shift, which performs sign-extension."

As discussed in the two's complement proposal, all known modern C++ compilers
already behave that way.  And it is unlikely any would go off and do something
different now, with C++20 tightening things up.

So I think relying on sign extension by right shift is fine.

kimbarrett replied:
That comment quoted earlier from globalDefinitions.hpp could be expanded to include the above analysis.

chhagedorn replied:
Thanks a lot Kim for your input and the detailed comments! I've included it as a comment at the `java_shift_right()` method and updated the code to directly use `java_shift_right()` instead of `shift_right_arithmetic()`.

merykitty reviewed (approved):
Otherwise looks good to me, thanks a lot

vnkozlov reviewed (approved):
Good.

chhagedorn commented:
Thanks Vladimir for approving it again. I've re-run some testing with latest master which looked good.

chhagedorn commented:
/integrate
